### PR TITLE
Gloomrot update

### PR DESCRIPTION
**Pull request by iZastic**
Updated to work with Gloomrot update.

Chat hook was re-written to manually process messages via OnUpdate because I was unable to get the hook into the ChatMessageJob OriginalLambdaBody to work. I think this is due to the ChatMessageJob being a struct now instead of a class.

### CODE REVIEW

File: Wetstone.csproj [hunk 1]
```diff
@@ -55,8 +53,8 @@
       <HintPath>$(UnhollowedDllPath)\com.stunlock.network.dll</HintPath>
     </Reference>
 
-    <Reference Include="UnityEngine.InputLegacyModule.dll">
-      <HintPath>$(UnhollowedDllPath)\UnityEngine.InputLegacyModule.dll</HintPath>
+    <Reference Include="Stunlock.Localization">
+      <HintPath>..\BepInEx\unhollowed\Stunlock.Localization.dll</HintPath>
```
Comment (decaprime) [inline Wetstone.csproj:57]: I'd prefer to use `<PackageReference Include="VRising.Unhollowed.Client" Version="0.6.0.57108*" />` over local refs.

File: Hooks/Chat.cs [hunk 2]
```diff
@@ -21,53 +18,51 @@ public static class Chat
...
+            var fromCharacter = VWorld.Server.EntityManager.GetComponentData<FromCharacter>(entity);
+            var ev = new VChatEvent(fromCharacter.User, fromCharacter.Character, chatMessage.MessageText.ToString(), chatMessage.MessageType);
+
+            WetstonePlugin.Logger.LogInfo($"[Chat] [{ev.Type}] {ev.User.CharacterName}: {ev.Message}");
+
+            try
+            {
+                OnChatMessage?.Invoke(ev);
+
+                if (ev.Cancelled) return;
```
Comment (decaprime) [inline Hooks/Chat.cs:57]: Does this work to cancel messages as intended? I think there may be something lost in the translation from the native version. If a downstream plugin marks the event as canceled the OnUpdate prefix exits processing the loop of entities, but the original chat message entity and it's processing seem unaffected. Also the query tends to have a single message, but in theory there could be more and then I think this just ignores those for a frame and then they could be unprocessed and exited by the next one. 

I'd not return here, then either 'not support' Cancelled anymore or simulate it for the callers.

You could considering using this whisper back from the VCF gloomrot changes to simulate canceling the message to the rest of the server: https://github.com/decaprime/VampireCommandFramework/commit/6871f4b00e3df1094856ff60817ed52b84723405#diff-4fca44e8b9150400cd4f27773e9541b0dfc90f2c50fab1819b960b5a7be6f896 .

File: Hooks/Chat.cs [hunk 3]
```diff
@@ -21,53 +18,51 @@ public static class Chat
     /// </summary>
     public static event ChatEventHandler? OnChatMessage;
```
Comment (decaprime) [inline Hooks/Chat.cs:19]: Might be good to add a remarks to explain the compromise for this hook, or maybe even a deprecation notice to this functionality. I understood why this hook existed given the native detour, but it's not as obvious to me outside of compatibility.

**Review by decaprime — COMMENTED** [review]
Thank you for working on this PR! @iZastic what tests/validations were done for this PR?

I'd consider tests surrounding
- CustomNetworkEvent
- Keybind: Register/Unregister and Load/Save
- Chat Event Cancel

File: Wetstone.csproj (hunk 1 above)
Comment (iZastic) [reply Wetstone.csproj:57]: That's not updated for the current hotfix patch number. Also, I didn't want to change the project, just update it. I guess I can do that though.

File: Hooks/Chat.cs (hunk 2 above)
Comment (iZastic) [reply Hooks/Chat.cs:57]: Hmm, i just moved the code to a new funciton. But the now that you mention it, the return doesn't make sense here.

File: Hooks/Chat.cs (hunk 3 above)
Comment (iZastic) [reply Hooks/Chat.cs:19]: Not sure what you're looking for here. I have a comment in the PR. The chat feature isn't deprecated, it's just changed internally.

**Comment by iZastic** [comment]
CustomNetworkEvent - I used RegisterBiDirectionalStruct to test the network serialization

Keybind: Register/Unregister and Load/Save - I have 2 keybinds registered and they are both working

Chat Event Cancel - Cancelation seems to be working. (using !reload, wetstone cancels the event and my test mod can see that it's canceld). Also, tested to see if destroyed entities are still retuned in the query after processing, and they are not.

File: Wetstone.csproj (hunk 1 above)
Comment (decaprime) [reply Wetstone.csproj]: Thanks for the reminder I've updated it on nuget.org: `<PackageReference Include="VRising.Unhollowed.Client" Version="0.6.1.*" />`

File: Hooks/Chat.cs (hunk 3 above)
Comment (decaprime) [reply Hooks/Chat.cs:19]: Never mind if only internal change. I read the implementation as intentionally not supporting the canceling in the way it use to work, resolving in favor of PR discussion of testing. The original thinking was that if mods depended on the message not going out in the original channel then they could be broken and it wouldn't just be an internal change.

**Comment by decaprime** [comment]
Have you tested this with multiple players? In particular the cancel for chat to prevent them going to the game. In lieu of not being able to stop them from going to the game, I suggest a hack to prevent *other* players from seeing the original message, I think that most consumers expect this more than the 'game' seeing it.

Chat has two expectations on cancellation in my interpretation of original implementation and docs:

> Chat messages will be forwarded to plugins first, and only after to the game, unless the event has been cancelled.
1. prevent sending to game

> Due to the way the cancellation system works, it is recommended that you check whether the event is cancelled before you process the message, and discard it if you do. Only process cancelled events if you have a good reason to do so.
2. you can check if an earlier plugin canceled the event

I think we still need to test 1, but you've already covered 2.

**Comment by iZastic** [comment]
The message entity is destroyed if it is canceled, and as mentioned above "Also, tested to see if destroyed entities are still returned in the query after processing, and they are not".

I have not tested with multiple players, I only have 1 client.

File: API/Keybindings.cs [hunk 4]
```diff
@@ -200,14 +199,14 @@ public Keybinding(KeybindingDescription description)
...
+        var hash = xxHash64.ComputeHash(idBytes, idBytes.Length);
         var invalid = false;
         do
         {
             invalid = false;
-            foreach (var entry in Enum.GetValues(typeof(InputFlag)).Cast<long>())
+            foreach (var entry in Enum.GetValues(typeof(InputFlag)))
             {
-                if (hash == entry)
+                if (hash == (ulong)entry.GetHashCode())
```
Comment (Kasuromi) [inline API/Keybindings.cs:209]: This doesn't have the same behavior, use the generic version of Enum.GetValues and cast to `ulong` for clarity.
```cs
foreach (var entry in Enum.GetValues<InputFlag>())
{
    if (hash == (ulong)entry)
    {
        invalid = true;
        hash -= 1;
    }
}
```

File: Util/NativeHookUtil.cs [hunk 5]
```diff
@@ -1,28 +1,48 @@
...
     }
 
-    public static FastNativeDetour Detour<T>(MethodInfo method, T to, out T original) where T : System.Delegate?
+    public static NativeDetour Detour<T>(MethodInfo method, T to, out T original) where T : System.Delegate?
+    {
+        var address = Il2CppMethodResolver.ResolveFromMethodInfo(method);
+        WetstonePlugin.Logger.LogInfo($"Detouring {method.DeclaringType?.FullName}.{method.Name} at {address.ToString("X")}");
+        var detour = new NativeDetour(address, Marshal.GetFunctionPointerForDelegate(to));
+        original = detour.GenerateTrampoline<T>();
+        return detour;
```
Comment (Kasuromi) [inline Util/NativeHookUtil.cs:29]: What's the motivation behind providing an API around MonoMod's `NativeDetour` and BepInEx's `INativeDetour` at the same time?

You should only provide API that returns INativeDetour.

File: Network/SerializationHooks.cs [hunk 6]
```diff
@@ -37,57 +37,52 @@ public static void Uninitialize()
...
     // --------------------------------------------------------------------------------------
 
     [UnmanagedFunctionPointer(CallingConvention.StdCall)]
-    public unsafe delegate void DeserializeEvent(IntPtr commandBuffer, IntPtr netBuffer, DeserializeNetworkEventParams eventParams);
+    public unsafe delegate void DeserializeEvent(IntPtr entityManager, IntPtr commandBuffer, IntPtr netBuffer, DeserializeNetworkEventParams eventParams);
 
     public static DeserializeEvent? DeserializeOriginal;
 
-    public unsafe static void DeserializeHook(IntPtr commandBuffer, IntPtr netBuffer, DeserializeNetworkEventParams eventParams)
+    public unsafe static void DeserializeHook(IntPtr entityManager, IntPtr commandBuffer, IntPtr netBuffer, DeserializeNetworkEventParams eventParams)
```
Comment (Kasuromi) [inline Network/SerializationHooks.cs:75]: Taking these parameters as native integers is redundant, you end up turning them into the relevant ref structs anyway. Take them in as ref parameters instead.
```suggestion
    public unsafe static void DeserializeHook(IntPtr entityManager, IntPtr commandBuffer, ref NetBufferIn netBuffer, DeserializeNetworkEventParams eventParams)
```

File: Network/SerializationHooks.cs [hunk 7]
```diff
@@ -37,57 +37,52 @@ public static void Uninitialize()
...
     {
-        // this is the same adjustment that we do in the Serialize hook
-        var netBufferIn = new NetBufferIn(new IntPtr((long)(netBuffer - 0x10)));
+        ref NetBufferIn netBufferIn = ref *(NetBufferIn*)netBuffer;
```
Comment (Kasuromi) [inline Network/SerializationHooks.cs:77]: Conversion is redundant, netBuffer should be taken in as a `ref NetBufferIn`

File: Network/SerializationHooks.cs [hunk 8]
```diff
@@ -37,57 +37,52 @@ public static void Uninitialize()
...
 
+
     [UnmanagedFunctionPointer(CallingConvention.StdCall)]
-    public unsafe delegate void SerializeEvent(IntPtr entityManager, UInt64 networkEventType, UInt64 netBufferOut, IntPtr entity);
+    public unsafe delegate void SerializeEvent(IntPtr entityManager, IntPtr networkEventType, IntPtr netBufferOut, IntPtr entity);
 
     public static SerializeEvent? SerializeOriginal;
 
-    public unsafe static void SerializeHook(IntPtr entityManager, UInt64 networkEventType, UInt64 netBufferOut, IntPtr entity)
+    public unsafe static void SerializeHook(IntPtr entityManager, IntPtr networkEventType, IntPtr netBuffer, IntPtr entity)
```
Comment (Kasuromi) [inline Network/SerializationHooks.cs:46]: Taking these parameters as native integers is redundant, you end up turning them into the relevant ref structs anyway. Take them in as ref parameters instead.

```suggestion
    public unsafe static void SerializeHook(IntPtr entityManager, IntPtr networkEventType, ref NetBufferOut netBuffer, ref Entity entity)
```

File: Network/SerializationHooks.cs [hunk 9]
```diff
@@ -37,57 +37,52 @@ public static void Uninitialize()
...
+            SerializeOriginal!(entityManager, networkEventType, netBuffer, entity);
             return;
         }
 
-        // we need to adjust by 0x10 here because the managed proxy that Il2CppUnhollower
-        // generates expects a boxed struct, whereas we have a pointer to an unboxed one.
-        // subtracting 0x10 is exactly two pointers, which allows us to fake a boxed struct
-        // with the same underlying data
-        var netBuffer = new NetBufferOut(new IntPtr((long)(netBufferOut - 0x10)));
+        ref NetBufferOut netBufferOut = ref *(NetBufferOut*)netBuffer;
```
Comment (Kasuromi) [inline Network/SerializationHooks.cs:57]: Conversion is redundant, netBuffer should be taken in as `ref NetBufferOut`

File: Hooks/Keybindings.cs [hunk 10]
```diff
@@ -36,7 +36,7 @@ static class Keybindings
 #nullable disable
     private static TryGetInputFlagLocalization_t TryGetInputFlagLocalization_Original;
     private static Harmony _harmony;
-    private static FastNativeDetour _detour;
+    private static NativeDetour _detour;
```
Comment (Kasuromi) [inline Hooks/Keybindings.cs:39]: Why MonoMod native detour? Does this even work? Use BepInEx's detouring API.
```suggestion
    private static INativeDetour _detour;
```

**Review by Kasuromi — CHANGES_REQUESTED** [review]
Minor suggestions on cleaning up the native detours so no redundant conversions are done.
Curious as to why MonoMod native detours were introduced into the codebase.

**Review by Kasuromi — APPROVED** [review]

File: Wetstone.csproj (hunk 1 above)
Comment (decaprime) [reply Wetstone.csproj]: This doesn't need to be in this PR, though I'd love for it to have that validation even right now, it can be done separately.

**Comment by decaprime** [comment]
> The message entity is destroyed if it is canceled, and as mentioned above "Also, tested to see if destroyed entities are still returned in the query after processing, and they are not".
> 
> I have not tested with multiple players, I only have 1 client.

Sorry, I don't think you need testing with multiple players anymore for chat event. I missed but now see the new Destroy call was added in https://github.com/molenzwiebel/Wetstone/commit/5d189814842e46adbf67eb4c98bee0adce1336a2 .

Great job @iZastic and also thank you for lending your expertise @Kasuromi !

**Pull request merged**